Create SelectPanel guidelines (#455)
primer/design: content/components/selectpanel.mdx

---

> title: SelctPanel

Reviewer: ```suggestion
title: Select panel
```

reactId: selectpanel
figma:

> description: SelectPanel is a semantic dialog that allows for complex selection options within an overlay.

Reviewer: ```suggestion
description: Select panel is a semantic dialog that allows for complex selection options within an overlay.
```

This is just a grammatical formatting decision I think we established at one point, so just suggesting changes to be consistent with that (not trying to be nitpicky!)

---

import {Box} from '@primer/react'
import ComponentLayout from '~/src/layouts/component-layout'
export default ComponentLayout

## Overview

> SelectPanel is used for overlay containers that contain multiple control sets which includes advanced single or multi-selection with a filter. SelectPanel can be anchored to a trigger, or centered within a view.
> 
> Commonly, SelectPanel is used for metadata application to an Issue or Pull Request, like assignee and label selections, or the branch selector on a repository’s code tab.

Reviewer: ```suggestion
Select panel is used for overlay containers that contain multiple control sets which includes advanced single or multi-selection with a filter. Select panel can be anchored to a trigger, or centered within a view.

Commonly, Select panel is used for metadata application to an Issue or Pull Request, like assignee and label selections, or the branch selector on a repository’s code tab.
```

## Anatomy

<img
  width="960"
  alt="Select panel for assigning labels to an issue. 'Apply labels to this issue' heading, close button in upper right hand corner. Search input with a magnifer icon as a visible label, listbox of label options and the first three are selected with a visual checkbox. The footer contains a button to edit labels, and cancel and apply buttons. Purple labels are overlayed onto Select panel: Overlay (dialog) pointing to the container, Search input pointing to the input, role='listbox' bracket of the listbox content, Optional action control to the button in the lower left corner, Explicit confirm / submission buttons bracket for the apply and cancel buttons in the lower right corner."
  src="https://user-images.githubusercontent.com/40274682/228329313-7c61fff5-0f2f-4170-b319-723e8a2a3a7d.png"
/>

## Options

> <img

Reviewer 2: I noticed the search icon in the `anatomy` image, but it's not included in the first `options` image. Not sure if that was deliberate, but it feels like it should be there!

Author: Good call!

  width="960"
  alt="First Select panel is a branch switcher control that includes tabs, a search input, a listbox with single-select items, and a footer with an All branches link, and cancel and apply buttons. Second Select panel is titled 'Show fields, with a list of about 10 multi-select items and apply and cancel buttons in the footer."

>   src="https://user-images.githubusercontent.com/40274682/228910129-c6b58046-4875-4aa7-afb4-f00384901ede.png"

Author: I would like to `selectpanel` to have the input as an optional prop. Like for the example on the right, even if it's just a list of multi-select options, it needs everything that a selectPanel has, minus the text input and I don't want to leave that up for devs to build each time they use it.

>   src="https://user-images.githubusercontent.com/40274682/228910129-c6b58046-4875-4aa7-afb4-f00384901ede.png"

Reviewer 3: Ok, then I think we should create an issue to support that prop in the PRC SelectPanel.

/>
<Caption>

>   Select panel can have as much as the branch selector component, or as little as just a multi-selection list. If the

Author: I can clarify this one more in the guidance. SelectPanel would still be user selection, but menu selection is purely referring to a semantic `<menu>`. If there is any thing in the popover in addition from a list of single selection, then it needs to become a popover, even if that single selection is combined with an input.

>   Select panel can have as much as the branch selector component, or as little as just a multi-selection list. If the

Author: Updated to: 

> Select panel can have as much as the branch selector component, or as little as just a multi-selection list. If the
  contents request user input beyond a semantic `<menu>`(example: the context menu on comments in GitHub), it’s likely you need a Select panel.

> <Caption>
>   SelectPanel can have as much as the branch selector component, or as little as just a multi-selection list. If the
>   contents request user input beyond menu selection, it’s likely you need a SelectPanel.
> </Caption>

Reviewer: ```suggestion
<Caption>
  Select panel can have as much as the branch selector component, or as little as just a multi-selection list. If the
  contents request user input beyond menu selection, it’s likely you need a Select panel.
</Caption>
```

### Additional control slot

<img
  width="960"
  alt="Three cropped Select panel footers each displaying a different option of a control in the bottom right left corner: Link, button, or checkbox input."

>   src="https://user-images.githubusercontent.com/40274682/228910127-b2c05a32-1eeb-4ebc-b5cc-6c089ecc7d2e.png"

Author: @hectahertz's latest version has the submit and cancel buttons. We would have to allow a prop for the third control that could be in the lower left hand corner. That's something we can create an issue for.

/>

> <Caption>
>   An optional control slot is available in the SelectPanel’s bottom left corner. Avoid use if not absolutely necessary
>   in order to avoid component and decision overwhelm.
> </Caption>

Reviewer: ```suggestion
<Caption>
  An optional control slot is available in the Select panel’s bottom left corner. Avoid use if not absolutely necessary
  in order to avoid component and decision overwhelm.
</Caption>
```

### Anchoring

> SelectPanel can be anchored to the trigger, or centered in the web page. If it’s centered in the web page it should have a backdrop.

Reviewer: ```suggestion
Select panel can be anchored to the trigger, or centered within the page. If it’s centered within the page it should have a backdrop.
```

> Select panel can be anchored to the trigger, or centered within the page. If it’s centered within the page it should have a backdrop.

Author: `SelectPanel` is a dialog always, but it's going to come with extra affordances so developers don't have to build it all from scratch each time with the proper behaviors and semantics. 

I think it's best left up to Product Designers for when it's anchored and when it's not as it will be heavily context based. When it's helpful to see the context of your screen I would anchor it to a trigger, when it's not as important or needs to be larger than ~320px, I would center it to the page.

> Select panel can be anchored to the trigger, or centered within the page. If it’s centered within the page it should have a backdrop.

Author: For now we can leave it as only anchored to a trigger and if we get feedback that they need it centered for a certain reason we can adjust. But they will all be modal dialogs.

<Box display="grid" gridTemplateColumns={['1fr', null, null, null, '1fr 1fr']} gridGap={5}>
  <div>
    <img
      src="https://user-images.githubusercontent.com/40274682/228386690-dd175879-0dd6-4798-8fb5-b0ee096eceea.png"
      alt=""
    />

>     <Caption>Anchored SelectPanel to a trigger.</Caption>

Reviewer: ```suggestion
    <Caption>Anchored Select panel to a trigger.</Caption>
```

  </div>
  <div>
    <img
      src="https://user-images.githubusercontent.com/40274682/228386689-ea191d76-3a50-41d9-98f0-9dc997caf4bf.png"
      alt=""
    />

>     <Caption>SelectPanel centered with a darker backdrop.</Caption>

Reviewer: ```suggestion
    <Caption>Select panel centered with a darker backdrop.</Caption>
```

  </div>
</Box>

## Best practices

> 

Reviewer 2: I think some of the imagery and captions are mixed up. I'm not sure if this is a formatting issue, or something else.

Reviewer 4: I noticed this too, but I think this may be a formatting issue, if you're using the rendered view withing GitHub.

Author: Yeah, I just verified in the code and it appropriately matches the way it should.

### Be predictable

<DoDontContainer>
  <Do>
    <img
      width="464"
      alt="SelectPanel with a back button, close button, Heading, subheading, five multiple choice options with three of them selected, and apply and cancel buttons."
      src="https://user-images.githubusercontent.com/40274682/228386683-20e7ee4e-0d1b-4b6c-9dd1-eddd277b5512.png"
    />
    <Caption>If there are more than single options in a menu, have a user confirm their choices.</Caption>
  </Do>
  <Dont>
    <img
      width="464"
      alt=""
      src="https://user-images.githubusercontent.com/40274682/228386685-4699374e-5f98-4dd4-bab8-f488ad0fff8a.png"
    />
    <Caption>
      Never force users to submit their multi-select options by closing the dialog or auto-updating their choices upon
      selection.
    </Caption>
  </Dont>
</DoDontContainer>

### Click-to-dismiss behavior

<DoDontContainer>
  <Do>
    <img
      width="464"
      alt="Mouse cursor indicating a clicking action on the apply button."
      src="https://user-images.githubusercontent.com/40274682/228388751-12c1f46e-aee7-4e03-8899-b273d31f7d1b.png"
    />
    <Caption>

>       Guide users to save or cancel their selections to dismiss SelectPanel to prevent confusion and data loss.

Reviewer: ```suggestion
      Guide users to save or cancel their selections to dismiss Select panel to prevent confusion and data loss.
```

    </Caption>
  </Do>
  <Dont>
    <img
      width="464"
      alt="Mouse cursor indicating a click action outside of the dialog."
      src="https://user-images.githubusercontent.com/40274682/228388749-445619e6-91fa-4f9f-b300-6e1fe21c62a5.png"
    />
    <Caption>

>       Don’t allow a click-outside-to-dismiss behavior for Select panel. Users may be used to closing and submitting the

Author: The escape key should always close and cancel changes, it should not save or retain any data. Places where this is in practice today is incorrect behavior. 

We can leave in the click to dismiss, but it _**cannot**_ submit any information. Whether or not it retains the information is potentially up for debate, I'd like to check with @jscholes on this. 

As for not seeing who is selected, labels and assignees that have already been applied should be at the top of the listbox, potentially in its own group. Selected items will not move to the top until it's been submitted and reopened.

>       Don’t allow a click-outside-to-dismiss behavior for Select panel. Users may be used to closing and submitting the

Reviewer 4: > As for not seeing who is selected, labels and assignees that have already been applied should be at the top of the listbox, potentially in its own group. Selected items will not move to the top until it's been submitted and reopened.

I wasn't clear with my example. Here's a screenshot of the current issue sidebar:

![A screenshot of the issues sidebar. In the screenshot the following headers are captured: Reviewers, Assignees, Labels. Reviewers has 4 reviewers listed, Assignees has none, and Labels has none. Each label has an icon button in the shape of a gear, which when acted upon will open the select panel respective to that heading.](https://user-images.githubusercontent.com/118266/228803054-3b066e02-e369-4608-850e-7fcbc96b9531.png)

If I click the gear to open the reviewers panel, the rest of the sidebar is occluded. Let's say I click the "Reviewers" gear to assign reviewers, but in that moment I realise I want to look at the Assignees list which will help me inform who I should add to reviewers. The only way to glance at the Assignees list is to dismiss the Reviewers panel, so I can check the Assignees list, then I can re-open to the Reviewers panel.

Keyboard users have common gesture available to them to dismiss floating UI; the `Escape` key. For any floating UI, _so far_, mouse users have a similar gesture, which is to click outside of the bounds of the floating UI. As is currently proposed, SelectPanel will deviate from that norm, and instead rely on the Cancel button, which is a smaller hit target.

> The escape key should always close and cancel changes, it should not save or retain any data. Places where this is in practice today is incorrect behavior.

I understand why it should not _submit_ any data, as it is potentially performing destructive actions. Can we please justify why it should not retain data during the page session. I don't understand why `Escape` would effectively reset the form. Is this a common pattern? Does `Escape` reset forms that are not inside floating UI?

>       Don’t allow a click-outside-to-dismiss behavior for Select panel. Users may be used to closing and submitting the

Author: > If I click the gear to open the reviewers panel, the rest of the sidebar is occluded. Let's say I click the "Reviewers" gear to assign reviewers, but in that moment I realise I want to look at the Assignees list which will help me inform who I should add to reviewers. The only way to glance at the Assignees list is to dismiss the Reviewers panel, so I can check the Assignees list, then I can re-open to the Reviewers panel.

If this becomes an issue for users with the new select panel, we can look into creating a minimizing or hiding behavior but as it's more complex I have not included this in the MVP which really needs to be published to improve this component that's been in limbo with me advocating for it for 2 years. 

> I don't understand why Escape would effectively reset the form. Is this a common pattern?

Yes this is the expected behavior of the escape key. It's the equivalent of "Get me out I don't want to be here". It should not save any information or changes made. 

We can allow click to dismiss if you feel this is necessary but it will result in data loss for users since GitHub has long since employed a anti-pattern. 

So we can either:
- Disable click-to-dismiss to prevent a loss of data entry
OR
- Enable click-to-dismiss and have a confirmation dialog pop up to prevent unsaved changes being lost (though this feels pretty heavy)

>       Don’t allow a click-outside-to-dismiss behavior for SelectPanel. Users may be used to closing and submitting the
>       information. Doing so would result in a loss of their changes.{' '}
>     </Caption>

Reviewer: ```suggestion
      Don’t allow a click-outside-to-dismiss behavior for Select panel. Users may be used to closing and submitting the
      information. Doing so would result in a loss of their changes.{' '}
    </Caption>
```

>       information. Doing so would result in a loss of their changes.{' '}

Author: Confirmed with @jscholes. Click to dismiss behavior should _not_ be different from pressing esc, and pressing esc needs to close and cancel changes.

  </Dont>
</DoDontContainer>

## Behavior

### Opening

> <img

Reviewer 2: Since you're referencing the `enter` key, I think it would make sense to replace `Return` with `Enter` in the image to align language.

Reviewer 4: I concur; `Enter` is the preferred key name here.

Author: Got it, I'll fix. We should update some Figma components then too.

  width="960"
  alt="Mouse cursor hovering over an icon button and a focused icon button with the space and return keys visually below the icon button."
  src="https://user-images.githubusercontent.com/40274682/228386681-e75b3bcd-ac0d-4e4a-afda-2a8dfc3a13a6.png"
/>
<Caption>

>   A SelectPanel trigger should be a semantic button. Mouse users can click and keyboard users can open with space or
>   enter.

Reviewer: ```suggestion
  A Select panel trigger should be a semantic button. Mouse users can click and keyboard users can open with space or
  enter.
```

</Caption>

<img
  width="960"
  alt="First select panel has focus applied to the search input with a blue 2px border. Second SelectPanel has no visible focus ring as focus is brought to the H1 of the dialog."
  src="https://user-images.githubusercontent.com/40274682/228386679-02d288f4-7e1c-4a77-a654-03def368e266.png"
/>
<Caption>

>   If SelectPanel has an input, focus should be placed within the input. If not, focus should be placed in the next
>   logical place. That next logical place might just be the H1 of the dialog.
> </Caption>

Reviewer: ```suggestion
  If Select panel has an input, focus should be placed within the input. If not, focus should be placed in the next
  logical place. That next logical place might just be the H1 of the dialog.
</Caption>
```

>   logical place. That next logical place might just be the H1 of the dialog.

Reviewer 4: This is a small nit but browsers have an algorithm for moving focus to new floating UI (dialog, popover) and they're unlikely to move it to an `h1`. We might want to adjust this to fall in-line with what the browser is more likely to do: focus on the `autofocus` item, or failing that first item that's a link, form control, or the first item with a non-negative `tabindex`.

Author: Yeah I'll revise this. In `SelectPanel`s case, we'll always have interactive elements as the main part of the dialog, so focus will go to the first interactive element. This is more for dialogs in general.

### Closing

<Box display="grid" gridTemplateColumns={['1fr', null, null, null, '1fr 1fr']} gridGap={5}>
  <div>
    <CustomVideoPlayer
      width="100%"
      src="https://user-images.githubusercontent.com/40274682/228390156-dab0b413-3b90-4736-a0e7-8d5d30665c43.mp4"
    />
    <Caption>
      Pressing esc, or selecting “cancel” should close the current dialog and return focus to the previous trigger. It

>       should not retain or submit any user input.

Reviewer 4: Highlighting this as something that should be discussed with the above comments around dismissal.

    </Caption>
  </div>
  <div>
    <CustomVideoPlayer
      width="100%"
      src="https://user-images.githubusercontent.com/40274682/228390297-d6d59818-1d1e-47bb-ac9d-cd1eccc57ba8.mp4"
    />
    <Caption>Selecting the submit button should submit any user input.</Caption>
  </div>
</Box>

### Responsive

> SelectPanel can become a full screen dialog or a bottom sheet at narrow viewports. Support at 320px is required if the SelectPanel is larger at wider viewports. See [narrow-viewport placement](https://primer.style/design/components/dialog#narrow-viewport-placement) in the Dialog guidelines for more details.
> 

Reviewer: ```suggestion
Select panel can become a full screen dialog or a bottom sheet at narrow viewports. Support at 320px is required if the Select panel is larger at wider viewports. See [narrow-viewport placement](https://primer.style/design/components/dialog#narrow-viewport-placement) in the Dialog guidelines for more details.

```

## Keyboard navigation

### Trigger

| Key                                                                                                                                                       | Description            |
| --------------------------------------------------------------------------------------------------------------------------------------------------------- | ---------------------- |

> | <Box as="kbd" backgroundColor="canvas.subtle" borderRadius={1} borderColor="border.default" borderWidth={1} borderStyle="solid" py={1} px={2}>Enter</Box> | Opens the SelectPanel. |

Reviewer: ```suggestion
| <Box as="kbd" backgroundColor="canvas.subtle" borderRadius={1} borderColor="border.default" borderWidth={1} borderStyle="solid" py={1} px={2}>Enter</Box> | Opens the Select panel. |
```

> | <Box as="kbd" backgroundColor="canvas.subtle" borderRadius={1} borderColor="border.default" borderWidth={1} borderStyle="solid" py={1} px={2}>Space</Box> | Opens the SelectPanel. |

Reviewer: ```suggestion
| <Box as="kbd" backgroundColor="canvas.subtle" borderRadius={1} borderColor="border.default" borderWidth={1} borderStyle="solid" py={1} px={2}>Space</Box> | Opens the Select panel. |
```

> ### SelectPanel

Reviewer: ```suggestion
### Select panel
```

| Key                                                                                                                                                             | Description                                                                                                                 |
| --------------------------------------------------------------------------------------------------------------------------------------------------------------- | --------------------------------------------------------------------------------------------------------------------------- |

> | <Box as="kbd" backgroundColor="canvas.subtle" borderRadius={1} borderColor="border.default" borderWidth={1} borderStyle="solid" py={1} px={2}>Enter</Box>       | Submits the information in the form when focused in the listbox. Activates buttons or links focused within the SelectPanel. |

Reviewer: ```suggestion
| <Box as="kbd" backgroundColor="canvas.subtle" borderRadius={1} borderColor="border.default" borderWidth={1} borderStyle="solid" py={1} px={2}>Enter</Box>       | Submits the information in the form when focused in the listbox. Activates buttons or links focused within the Select panel. |
```

> | <Box as="kbd" backgroundColor="canvas.subtle" borderRadius={1} borderColor="border.default" borderWidth={1} borderStyle="solid" py={1} px={2}>Enter</Box>                                    | Submits the information in the form when focused in the listbox. Activates buttons or links focused within the Select panel. |

Reviewer 4: It might be good to call out what `Enter` should do when the Search Input is focussed, even if that is "it should do nothing", otherwise it could be left to interpretation.

> | <Box as="kbd" backgroundColor="canvas.subtle" borderRadius={1} borderColor="border.default" borderWidth={1} borderStyle="solid" py={1} px={2}>Tab</Box>                                      | Move focus forward between interactive controls and widgets within the dialog.                                              |

Reviewer 4: When the last control is focussed, does the `Tab` key cycle to the top of the controls within the select panel? In other words - are we focus trapping?

If we are focus trapping, does that give users who cannot see the page the same experience as those who can? If a sighted user is able to see the rest of the contents of the page, but a user who is reliant on moving focus is trapped within the bounds of the SelectPanel, that feels inconsistent to me.

Author: Yes, we need to be trapping focus. We need a solid boundary for these complex dialogs or else screen reader users will not know they've left the dialog and they will have to use context clues to realize they've left it. It could be a big problem if they perform a command thinking that they're still within the bounds of the dialog but they are not.

| <Box as="kbd" backgroundColor="canvas.subtle" borderRadius={1} borderColor="border.default" borderWidth={1} borderStyle="solid" sx={{ whiteSpace: "nowrap"}} py={1} px={2}>Shift + Tab</Box> | Move focus backward between interactive controls and widgets within the dialog.                                             |

> | <Box as="kbd" backgroundColor="canvas.subtle" borderRadius={1} borderColor="border.default" borderWidth={1} borderStyle="solid" py={1} px={2}>ArrowDown</Box>                                | Move focus from the input to the listbox. (Shift+Tab required to move from listbox to input)                                |

Reviewer 4: Is it worth adding `ArrowUp` here too, or do we omit that as convention?

Author: I could add `ArrowUp` for posterity. If you're at the first item in the listbox, nothing will happen, but if you're in any other item in the listbox, it will move to the item previous.

| <Box as="kbd" backgroundColor="canvas.subtle" borderRadius={1} borderColor="border.default" borderWidth={1} borderStyle="solid" py={1} px={2}>PageDown</Box>                                 | Move focus to last item in listbox.                                                                                         |
| <Box as="kbd" backgroundColor="canvas.subtle" borderRadius={1} borderColor="border.default" borderWidth={1} borderStyle="solid" py={1} px={2}>Esc</Box>                                      | Close dialog and clear any changed inputs.                                                                                  |

### Multi-select

| Key                                                                                                                                                           | Description                         |
| ------------------------------------------------------------------------------------------------------------------------------------------------------------- | ----------------------------------- |

> | <Box as="kbd" backgroundColor="canvas.subtle" borderRadius={1} borderColor="border.default" borderWidth={1} borderStyle="solid" py={1} px={2}>ArrowDown</Box> | Navigate through the listbox items. |

Reviewer 4: Is it worth adding `ArrowUp` here too, or do we omit that as convention?

Author: I could add ArrowUp for posterity. If you're at the first item in the listbox, nothing will happen, but if you're in any other item in the listbox, it will move to the item previous.

Author: I'm trying to keep the Multi-select and Single-select sections specific to just their different behaviors and any shared behaviors to be within Select panel. If that's confusing I could just remove that Select panel section and duplicate info

> | <Box as="kbd" backgroundColor="canvas.subtle" borderRadius={1} borderColor="border.default" borderWidth={1} borderStyle="solid" py={1} px={2}>Space</Box>     | Select an item in the listbox.      |

Reviewer 4: Is it worth adding `Enter` here to clarify the behaviour more precisely?

### Single-select

| Key                                                                                                                                                           | Description                                              |
| ------------------------------------------------------------------------------------------------------------------------------------------------------------- | -------------------------------------------------------- |

> | <Box as="kbd" backgroundColor="canvas.subtle" borderRadius={1} borderColor="border.default" borderWidth={1} borderStyle="solid" py={1} px={2}>ArrowDown</Box> | Navigate through listbox items. Selection follows focus. |

Reviewer 4: Is it worth adding `Enter` here to clarify the behaviour more precisely?

#### Example behavior

<Box display="grid" gridTemplateColumns={['1fr', null, null, null, '1fr 1fr']} gridGap={5}>
  <div>
    <CustomVideoPlayer
      width="100%"
      src="https://user-images.githubusercontent.com/40274682/228391029-44ab995b-7a04-4b85-825e-dc77f90838dd.mp4"
    />
    <Caption>

>       Keyboard interaction with a single-select SelectPanel demonstrating selection following focus with navigation via

Reviewer: ```suggestion
      Keyboard interaction with a single-select Select panel demonstrating selection following focus with navigation via
```

      the ArrowDown key.
    </Caption>
  </div>

> </Box>

Reviewer 3: It would be nice to add a "Related links" section where we link to similar components and related UI pattern guidelines.

Here's what come to mind:
```
- [Data saving guidelines](/ui-patterns/saving)
- [ActionMenu](/components/ActionMenu)
- [Select](/components/Select)
```
